design-history/exploration-reports/2019.06-unixfsv2-spike-01.md

```diff
@@ -0,0 +1,296 @@
+A unixfsv2 spike, using schemas
+===============================
+
+This document is an exploration of what a hypothetical "unixfs v2" might look like,
+using schemas as a design language for describing the data structures.
+
+This is not a final proposal, but rather a lens to explore both unixfsv2
+concepts and schemas themselves at the same time, and how we expect them
+to contribute to each other.  Several concepts in schemas, advanced layouts,
+their interactions, and their syntaxes will be trialled here; they are not final.
+
+Without further ado...
+
+Schema
+------
+
+```
+type AnyFile union {
+	| File "f"
+	| Dir "d"
+	| Symlink "s"
+} representation keyed
+
+type Filename string
+
+type Dir struct {
+	members {Filename:DirEnt}<HAMT>
+}
+
+type DirEnt struct {
+	content &AnyFile
+	attribs Attribs
+}
```

Comment:
This is the most repugnant comment I am going to leave here, but consider it in terms of pretty significant space-savings. Especially interesting if we agree to **not** use HAMT for a predefined raw-block size, as part of the spec. But even with HAMT the approach can still be applied and drastically reduce directory block sizes ( often by half )

Originally I took a very similar approach to yours - a decorator per link, but later @olcbean pointed out that within a typical distribution tarfile the value of `attribs` will be usually identical for all files ( `root/root/0644/identical-mtime` ) and for all directories ( `root/root/0755/identical-mtime` ).

Thus ( again, in my little fork-world of legacy protobuf ) I went with:

```
message Data {
	enum DataType {
		Raw = 0;
		Directory = 1;
		File = 2;
		Metadata = 3;
		Symlink = 4;
		HAMTShard = 5;
	}

	required DataType Type = 1;
	optional bytes Data = 2;
	optional uint64 filesize = 3;
	repeated uint64 blocksizes = 4;

	optional uint64 hashType = 5;
	optional uint64 fanout = 6;

	// contains interesting parts of the POSIX stat struct
	repeated DecoratorPOSIX DecoratorPOSIX = 7;
}

message DecoratorPOSIX {

        // 0-indexed array indicating which PBLink's does this decorator apply to
        // similar in spirit to blocksizes = 4
        // Mad props to @olcbean for pointing out that within an e.g. tar archive
        // almost all the modes/mtimes will converge to 2 or 3 unique decorators
        repeated uint32 DecoratedPBLinkIndices = 1 [packed=true];

        // Carries precise layout of https://golang.org/pkg/os/#FileMode
        // NOTE: on the wire this value will be *little* endian
        // Mode bit-positions are protocol-safe:
        // https://github.com/golang/go/issues/25422#issuecomment-389497051
        optional fixed32 FileMode = 2;

        // Each timespec weighs at 11 + 2 bytes of wrapping
        optional Timespec MTime = 3;
}


```